skopt/space/space.py

```diff
@@ -84,11 +90,44 @@ def check_dimension(dimension, transform=None):
     dimension : Dimension
         Dimension instance.
     """
+    old_dim = _check_dimension_old(dimension, transform=transform)
+    try:
+        new_dim = _check_dimension(dimension, transform=transform)
+    except Exception as err:
+        new_dim = f"<{err.__class__.__name__}: {err}>"
+    if new_dim != old_dim:
+        warnings.warn(f"Dimension {dimension} was inferred to {old_dim}. In "
+                      "upcoming versions of scikit-optimize, it will be "
+                      f"inferred to {new_dim}. See the documentation of the "
+                      "check_dimension function for the upcoming API.")
+    return old_dim
+
+
+def _check_dimension(dimension, transform=None):
+    if isinstance(dimension, Dimension):
+        return dimension
+    elif isinstance(dimension, (list, np.ndarray)):
+        return Categorical(dimension, transform=transform)
+    elif isinstance(dimension, tuple) and 1 <= len(dimension) <= 4:
+        low, high, *args = dimension
+        if (isinstance(low, numbers.Integral)
+                and isinstance(high, numbers.Integral)):
+            return Integer(int(low), int(high), *args, transform=transform)
+        elif isinstance(low, numbers.Real) and isinstance(high, numbers.Real):
+            return Real(float(low), float(high), *args, transform=transform)
```

Comment:
`low` and `high` are already `int` (or `float`). Why cast?

Reply:
The `float` cast is because one of the boundaries might be an `int`, and a `Real` dimension asks for two `float` in its doc.
The `int` cast was to support other integral types like `bool` or `np.int32`.

Since `int` is not a sublcass of `float`, I think the `float` cast is necessary to respect `Real`'s interface. I'm not too sure about the `int` cast, I mostly did it for symmetry.